### PR TITLE
Map refactor from NSF

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Cherry picked out of the nsf to reduce the size of the PR.

### CODE REVIEW

File: src/openrct2/rct2/S6Exporter.cpp [hunk 1]
```diff
@@ -194,8 +194,7 @@ void S6Exporter::Export()
     _s6.scenario_srand_0 = state.s0;
     _s6.scenario_srand_1 = state.s1;
 
-    // Map elements must be reorganised prior to saving otherwise save may be invalid
```
Comment (Gymnasiast) [inline src/openrct2/rct2/S6Exporter.cpp:197]: Could you restore this comment?

File: src/openrct2/world/Map.cpp [hunk 2]
```diff
@@ -1500,12 +1522,19 @@ void map_update_tiles()
...
 
-        auto mapPos = TileCoordsXY{ x, y }.ToCoordsXY();
-        auto* surfaceElement = map_get_surface_element_at(mapPos);
-        if (surfaceElement != nullptr)
+        // Repeat for each 256x256 block on the map
+        for (int32_t blockY = 0; blockY < gMapSize; blockY += 256)
         {
-            surfaceElement->UpdateGrassLength(mapPos);
-            scenery_update_tile(mapPos);
+            for (int32_t blockX = 0; blockX < gMapSize; blockX += 256)
```
Comment (Gymnasiast) [inline src/openrct2/world/Map.cpp:1528]: This looks wrong. BlockX seems to contain big coords - you check against gMapSize and add 256 on every iteration -, yet the value is used to create a TIleCoordsXY.

Comment (duncanspumpkin) [reply src/openrct2/world/Map.cpp:1528]: it isn't used until the nsf increases the map size. In normal map size it only ever does the one iteration.

File: src/openrct2/world/Map.cpp [hunk 3]
```diff
@@ -1111,21 +1139,16 @@ bool map_check_free_elements_and_reorganise(int32_t numElements)
...
-    {
-        log_error("Cannot insert new element");
-        return nullptr;
-    }
 
-    newTileElement = gNextFreeTileElement;
-    originalTileElement = gTileElementTilePointers[tileLoc.y * MAXIMUM_MAP_SIZE_TECHNICAL + tileLoc.x];
+    auto numElementsOnTileOld = CountElementsOnTile(loc);
+    auto numElementsOnTileNew = numElementsOnTileOld + 1;
+    auto* newTileElement = AllocateTileElements(numElementsOnTileNew);
```
Comment (duncanspumpkin) [inline src/openrct2/world/Map.cpp:1145]: Allocate can fail. Must return nullptr if fail.

File: src/openrct2/actions/PlacePeepSpawnAction.cpp [hunk 4]
```diff
@@ -47,12 +47,6 @@ GameActions::Result::Ptr PlacePeepSpawnAction::Query() const
     res->Expenditure = ExpenditureType::LandPurchase;
     res->Position = _location;
 
-    if (!map_check_free_elements_and_reorganise(3))
```
Comment (duncanspumpkin) [inline src/openrct2/actions/PlacePeepSpawnAction.cpp:50]: peep spawns dont use tile elements so this check was pointless. Likely copied from the park entrance code.

**Comment by Gymnasiast** [comment]
You merged this without any approval. That’s bad enough, but you also caused this problem: https://github.com/OpenRCT2/OpenRCT2/issues/14848

File: src/openrct2/world/TileInspector.cpp [hunk 5]
```diff
@@ -94,7 +94,7 @@ namespace OpenRCT2::TileInspector
     GameActionResultPtr InsertCorruptElementAt(const CoordsXY& loc, int16_t elementIndex, bool isExecuting)
     {
         // Make sure there is enough space for the new element
-        if (!map_check_free_elements_and_reorganise(1))
+        if (!!MapCheckCapacityAndReorganise(loc))
```
Comment (Basssiiie) [inline src/openrct2/world/TileInspector.cpp:97]: Double exclamation mark is probably not what you intended?

File: src/openrct2/world/TileInspector.cpp [hunk 6]
```diff
@@ -359,7 +359,7 @@ namespace OpenRCT2::TileInspector
     GameActionResultPtr PasteElementAt(const CoordsXY& loc, TileElement element, bool isExecuting)
     {
         // Make sure there is enough space for the new element
-        if (!map_check_free_elements_and_reorganise(1))
+        if (!!MapCheckCapacityAndReorganise(loc))
```
Comment (Basssiiie) [inline src/openrct2/world/TileInspector.cpp:362]: Same here.

**Pull request merged**